File: services/py/markdown_graph/graph.py
import os
import re
import sqlite3
from collections import deque
from typing import Iterable, List


LINK_PATTERN = re.compile(r"\[[^\]]*\]\(([^)]+\.md)\)")
TAG_PATTERN = re.compile(r"(?<!\w)#(\w+)")


class GraphDB:
    """Maintain links and hashtags for markdown files in SQLite."""

    def __init__(self, db_path: str, repo_path: str) -> None:
        self.repo_path = os.path.abspath(repo_path)
        self.conn = sqlite3.connect(db_path, check_same_thread=False)
        self._init_db()

    def _init_db(self) -> None:
        cur = self.conn.cursor()
        cur.execute(
            """
            CREATE TABLE IF NOT EXISTS files (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                path TEXT UNIQUE
            )
            """
        )
        cur.execute(
            """
            CREATE TABLE IF NOT EXISTS links (
                src INTEGER,
                dst INTEGER,
                FOREIGN KEY(src) REFERENCES files(id),
                FOREIGN KEY(dst) REFERENCES files(id)
            )
            """
        )
        cur.execute(
            """
            CREATE TABLE IF NOT EXISTS hashtags (
                file_id INTEGER,
                tag TEXT,
                FOREIGN KEY(file_id) REFERENCES files(id)
            )
            """
        )
        self.conn.commit()

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------
    def _rel(self, path: str) -> str:
        abs_path = os.path.abspath(
            path if os.path.isabs(path) else os.path.join(self.repo_path, path)
        )
        return os.path.relpath(abs_path, self.repo_path)

    def _get_file_id(self, path: str) -> int:
        rel = self._rel(path)
        cur = self.conn.cursor()
        cur.execute("INSERT OR IGNORE INTO files(path) VALUES (?)", (rel,))
        self.conn.commit()
        cur.execute("SELECT id FROM files WHERE path=?", (rel,))
        row = cur.fetchone()
        assert row is not None
        return row[0]

    def _parse_links(self, content: str) -> Iterable[str]:
        return LINK_PATTERN.findall(content)

    def _parse_tags(self, content: str) -> Iterable[str]:
        return TAG_PATTERN.findall(content)

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------
    def update_file(self, path: str, content: str) -> None:
        file_id = self._get_file_id(path)
        cur = self.conn.cursor()
        cur.execute("DELETE FROM links WHERE src=?", (file_id,))
        cur.execute("DELETE FROM hashtags WHERE file_id=?", (file_id,))
        for link in self._parse_links(content):
            target_path = os.path.normpath(os.path.join(os.path.dirname(path), link))
            target_id = self._get_file_id(target_path)
            cur.execute(
                "INSERT INTO links(src, dst) VALUES (?, ?)", (file_id, target_id)
            )
        for tag in self._parse_tags(content):
            cur.execute(
                "INSERT INTO hashtags(file_id, tag) VALUES (?, ?)", (file_id, tag)
            )
        self.conn.commit()

    def get_links(self, path: str) -> List[str]:
        rel = self._rel(path)
        cur = self.conn.cursor()
        cur.execute(
            """
            SELECT f2.path FROM links
            JOIN files f1 ON f1.id = links.src
            JOIN files f2 ON f2.id = links.dst
            WHERE f1.path=?
            """,
            (rel,),
        )
        return [row[0] for row in cur.fetchall()]

    def get_files_with_tag(self, tag: str) -> List[str]:
        cur = self.conn.cursor()
        cur.execute(
            """
            SELECT files.path FROM hashtags
            JOIN files ON files.id = hashtags.file_id
            WHERE tag=?
            """,
            (tag,),
        )
        return [row[0] for row in cur.fetchall()]

    def cold_start(self) -> None:
        start = os.path.join(self.repo_path, "readme.md")
        queue: deque[str] = deque([start])
        seen = set()
        while queue:
            path = queue.popleft()
            if path in seen or not os.path.isfile(path):
                continue
            seen.add(path)
            with open(path, "r", encoding="utf-8") as f:
                content = f.read()
            self.update_file(path, content)
            for link in self._parse_links(content):
                next_path = os.path.normpath(os.path.join(os.path.dirname(path), link))
                if next_path.startswith(self.repo_path):
                    queue.append(next_path)

File: services/ts/markdown-graph/src/graph.ts
import path from 'path';
import { promises as fs } from 'fs';
import { ContextStore, DualStoreManager } from '@shared/ts/dist/persistence/index.js';

const LINK_PATTERN = /\[[^\]]*\]\(([^)]+\.md)\)/g;
const TAG_PATTERN = /(?<!\w)#(\w+)/g;

export class GraphDB {
    private repoPath: string;
    private links: DualStoreManager<'text', 'timestamp'>;
    private hashtags: DualStoreManager<'text', 'timestamp'>;

    private constructor(
        repoPath: string,
        links: DualStoreManager<'text', 'timestamp'>,
        hashtags: DualStoreManager<'text', 'timestamp'>,
    ) {
        this.repoPath = path.resolve(repoPath);
        this.links = links;
        this.hashtags = hashtags;
    }

    static async create(repoPath: string): Promise<GraphDB> {
        const store = new ContextStore();
        const links = (await store.createCollection(
            'markdown_graph_links',
            'text',
            'timestamp',
        )) as DualStoreManager<'text', 'timestamp'>;
        const hashtags = (await store.createCollection(
            'markdown_graph_hashtags',
            'text',
            'timestamp',
        )) as DualStoreManager<'text', 'timestamp'>;
        return new GraphDB(repoPath, links, hashtags);
    }

    private rel(p: string): string {
        const abs = path.isAbsolute(p) ? p : path.join(this.repoPath, p);
        return path.relative(this.repoPath, path.resolve(abs));
    }

    private parseLinks(content: string): string[] {
        const results: string[] = [];
        for (const match of content.matchAll(LINK_PATTERN)) {
            results.push(match[1]!);
        }
        return results;
    }

    private parseTags(content: string): string[] {
        const results: string[] = [];
        for (const match of content.matchAll(TAG_PATTERN)) {
            results.push(match[1]!);
        }
        return results;
    }

    async updateFile(filePath: string, content: string): Promise<void> {
        const rel = this.rel(filePath);
        await this.links.mongoCollection.deleteMany({ 'metadata.src': rel } as any);
        await this.hashtags.mongoCollection.deleteMany({ 'metadata.path': rel } as any);

        const dir = path.dirname(rel);
        for (const link of this.parseLinks(content)) {
            const target = path.normalize(path.join(dir, link));
            const dst = this.rel(target);
            await this.links.mongoCollection.insertOne({
                text: dst,
                timestamp: Date.now(),
                metadata: { src: rel, dst },
            } as any);
        }
        for (const tag of this.parseTags(content)) {
            await this.hashtags.mongoCollection.insertOne({
                text: tag,
                timestamp: Date.now(),
                metadata: { path: rel, tag },
            } as any);
        }
    }

    async getLinks(p: string): Promise<string[]> {
        const rel = this.rel(p);
        const docs = await this.links.mongoCollection
            .find({ 'metadata.src': rel } as any)
            .toArray();
        return docs.map((d) => d.metadata?.dst as string);
    }

    async getFilesWithTag(tag: string): Promise<string[]> {
        const docs = await this.hashtags.mongoCollection
            .find({ 'metadata.tag': tag } as any)
            .toArray();
        return docs.map((d) => d.metadata?.path as string);
    }

    async coldStart(): Promise<void> {
        const start = path.join(this.repoPath, 'readme.md');
        const queue: string[] = [start];
        const seen = new Set<string>();

        while (queue.length) {
            const current = queue.shift()!;
            const rel = this.rel(current);
            if (seen.has(rel)) continue;
            seen.add(rel);
            try {
                const content = await fs.readFile(current, 'utf8');
                await this.updateFile(current, content);
                for (const link of this.parseLinks(content)) {
                    const next = path.normalize(path.join(path.dirname(current), link));
                    if (next.startsWith(this.repoPath)) {
                        queue.push(next);
                    }
                }
            } catch {
                // ignore missing files
            }
        }
    }
}
